### DISCUSSION TITLE
Port2 & request.

### DISCUSSION THREAD

**Discussion by AndyMcProducer** [jamulussoftware/jamulus#2205] [Ideas]
Hiya all, I see the info about the port2, has anyone used javascript only to pull and push to a server yet?
PHP is great but would be nice using it via js to as not all have php server or can use one.
I can but I'd much prefer to just pull and push via js.

Andre has his chat and server info on his site and it uses some js too but looking at it it seems maybe it's using
php behind it.
But would be a nice option for all to have in Jamulus, webchat and stream with out need to have a reconnecting client to stream.
Then maybe like with ninjam we can listen in before joining as not all jams are for everyone, but some just want to listen and hangout
to so the webchat and stream would be handy.
The server has record, can this be added too, to allow a stream out as well?
If not a shoutcast or icecast  stream then the audio that's being recorded to file output to an audio device as well.
Where this would be a stream output latency wouldn't be an issue and if all client audio had an limiter before mixed down to stereo
it would help keep levels nice for listeners. With a simple setting of -db for all clients.
Imo audio out from the server is better than to sc or ic as it then allows people to use and push however they like.
In server app could have drop down list box with audio devices to use then peeps could use a vad of their choosing.

Same for chat, if it could be pushed simple to a irc server and chan of choice this coudl be reused how peeps liked without
the need for many formats and such.
So someone could make an app to convert to a discord chan or use a private irc chan or run there own lil irc server.
It allows people to become more part of it all and achieve cool stuff with not needing to know how to program Jumulus.

CSS has grown and become very powerful now and can achieve a lot, so if jamulus ever was CSS then would be some
awesome themes to be made with some cool fx's and all sorts.
CSS can do a lot and until you see some of the cooler things done by CSS for most they wouldn't realize its potential.
It's just a much simpler way to handle and update a screen with options to have more skins and themes which in turn brings in more people.

Some nice css examples on here, some with code to.
https://www.mockplus.com/blog/post/css-animation-examples

**Comment by dcorson-ticino-com** [comment]
Isn't Javascript an interpreted language?
Slow.....

**Reply by AndyMcProducer** [reply]
You'd have to be on a slow system, it's used already on most things. Andre i think is using php but is using js to push/pull to the page.

**Comment by ann0see** [comment]
I think some of this functionality was already developed: https://github.com/jamulussoftware/jamulus/pull/1975 and the streamer2 PR.

@dtinth

**Reply by AndyMcProducer** [reply]
@ann0see Have the updates from @dtinth been merged with the master now? It looks like it has from the commits but in the title says like they want to be merged.

**Reply by ann0see** [reply]
No, unfortunately not since they aren't finished yet, I think. There are still some bugs.

**Reply by AndyMcProducer** [reply]
I see you did an update today to the streamer2, hows that going to work when finished, will it be to ice cast and shoutcast? Could it also be as audio output?
I guess too based on the instrument selection some high and low pass could be done to to improve on a final mix.
Just idea's for the future.
Could even be an on of option on a client side, this is a reasonable guide to frequencies.
Cuts up to the start of the fundamentals could be done and the harmonics could be cut/lowered if they breached so much db. and then the cuts after, would help peeps not ear kill others and allow for better overall mixed sound of all.
![image](https://user-images.githubusercontent.com/8242017/147887114-d7d1cb28-aaac-49cb-819e-7081b91dd7d3.png)

**Reply by ann0see** [reply]
> I see you did an update today to the streamer2

Yes. I pushed it to the main repo as [feature branch](https://github.com/jamulussoftware/jamulus/tree/feature/ffmpeg-streamer2) and resolved some merge conflicts. However I didn't test it since I don't run icecast. But I do think it can be used as stereo output (note n-channel output while n = number of people connected is not really possible due to the implementation, I think)? @dingodoppelt should know more

> I guess too based on the instrument selection some high and low pass could be done to to improve on a final mix.

Sure. But this could also be done now if you connect Jamulus via JACK to a DAW; But honestly, I doubt we could really put much FX into Jamulus - if any due to the KISS principle. Audio processing would need to happen outside Jamulus.

**Reply by AndyMcProducer** [reply]
using jack version allow multiple outs for clients?

**Reply by ann0see** [reply]
No. It's impossible to have more than 2 outputs for one client. And work on that has been done, but is stale: https://github.com/jamulussoftware/jamulus/pull/1913

**Comment by AndyMcProducer** [comment]
Sorry I meant per client output multi channel not multi channel per client. But Opus does support multichannel per person into 1 stream to.
Jesse used it in Sonobus to do multi channel per client.
It didn't take him that long to implement it but looking at the info about it it wasn't a big switch from orig method of using opus codec.
Opus supports multi channel up to 255 input/outputs per 1 stream.
https://opus-codec.org/docs/opus_api-1.3.1/
Opus: Opus - Opus Codec<https://opus-codec.org/docs/opus_api-1.3.1/>
The Opus codec is designed for interactive speech and audio transmission over the Internet. It is designed by the IETF Codec Working Group and incorporates technology from Skype's SILK codec and Xiph.Org's CELT codec.
opus-codec.org

________________________________

**Comment by AndyMcProducer** [comment]
https://opus-codec.org/docs/opus_api-1.3.1/group__opus__multistream.html
Opus: Opus Multistream API<https://opus-codec.org/docs/opus_api-1.3.1/group__opus__multistream.html>
The multistream API allows individual Opus streams to be combined into a single packet, enabling support for up to 255 channels. Unlike an elementary Opus stream, the encoder and decoder must negotiate the channel configuration before the decoder can successfully interpret the data in the packets produced by the encoder.
opus-codec.org

I wonder if the server packed all into 1 stream if it would bring down data and turn around. So like if the server put all clients into one multistream to a client. Maybe it is done like this already but if not slots could be pre attained then just filled with clients as they come and go.
________________________________
